### DISCUSSION TITLE
Show a notification when executing shell commands

### DISCUSSION THREAD

**Discussion by Taitava** [Taitava/obsidian-shellcommands#260] [Ideas - done & released]
A suggestion from @pulinagrawal in #259:
> As a compromise can we have something like a notification when a command starts?

My response:
> I think some kind of ability to monitor executed shell command processes could be planned. A notification balloon could be one approach, yes. Thank you for the idea! 🙂 It could later be developed further to make it more advanced.

---

I've now implemented this and here's how it looks like:

![kuva](https://user-images.githubusercontent.com/13002142/186178845-161898ca-c5c3-4ce9-9bf7-fe816acffdca.png)

Here it shows that the shell command with alias _A long running shell command_ is being executed now. If a shell command does not have an alias text defined, then the actual command is shown in the notification.

---

![kuva](https://user-images.githubusercontent.com/13002142/186178651-de4a46a4-7d2e-40bc-9fcf-d63a478f0a07.png)

A new notification setting is created in the _Output_ tab in the plugin's main settings. The setting affects all shell commands, there is no shell command specific setting for this.
- _Do not show_: This is selected by default.
- _Show for **n** seconds_: Shows the notification every time. The duration is the same as defined in _Notification message duration_.
- _Show until the process is finished_: Shows the notification every time, and keeps it visible for as long as the shell command runs. This is handy if you have long running shell commands without any visible output, and you want to keep an eye out on if they are still running or not. If you execute quickly running shell commands, then the notification will just show up and disappear immediately.
- _Show only if executing takes long_: Same as _Show until the process is finished_, but waits two seconds after starting the execution. If the shell command is still running after two seconds, the notification is then displayed.

---

@pulinagrawal, what do you think about the options? 🙂

**Comment by pulinagrawal** [comment]
@Taitava That looks just perfect. Thanks.

**Comment by Taitava** [comment]
# Released now
@pulinagrawal can you test this and let me know if there's anything to improve, please? 🙂

**Reply by pulinagrawal** [reply]
It works great. I like the option to show the notification until process is finished. I am gonna be using that option. @Taitava  Thanks for implementing it.

**Reply by Taitava** [reply]
Thank you for your feedback! ❤️

**Comment by snail123815** [comment]
Hi @Taitava,

Thanks for the new function, I like it. As I am not testing it yet, I got a question just by reading the screen shots.

Will the notification (Show until the process is finished) block all operation? I would like to have that ability if possible.

The reason is that I am implementing a command which contains `git add . && git commit -m "dum" && git push` etc. Although changing the file during this process is not a problem for execution, but it is still good that the input can be completely blocked while running this command.

I am sure for some other command which include variables like `{{caret_position}}`, the file change during execution might cause severe problems. Thus blocking all operations might be a good feature.

I am happy to help you test this.

Best~

**Reply by snail123815** [reply]
I just noticed the above function is implemented as a 'global' settings. Might be good to have it per command.

**Reply by Taitava** [reply]
Hi @snail123815 and thankyou for your ideas! 🙂 

## Blocking input is not supported
If I understood correctly, by _blocking input_ you mean to prevent the user doing anything in Obsidian, e.g. prevent editing files and prevent moving the text caret in the editor etc. The SC plugin does not do this. I assume implementing something like this reliably would be quite a big work and I'm not sure is it's officially supported by Obsidian API. If I remember correctly, Obsidian does allow plugins to use `Code Mirror`'s API (CM is the actual editor component that Obsidian uses). Perhaps Code Mirror's API supports making the editor readonly (hopefully preventing caret moving, too)? But I'm not sure, just guessing.

**At this point I'd say no to implementing blocking**. The work needed for this feels quite big compared to the benefits, and I'm not certain about the quality of the result. Maybe there could be an option to make the notification to be a modal instead. It would "distract" the user from doing any changes in Obsidian as long as the modal stays open. Of course the user can close the modal whenever they want, but at least they'll do it on purpose, not by mistake.

## Defining the setting on a per shell command basis

Yes, could be a good idea to be able to define this on a per shell command basis. The global setting could still be there to act as a default if a shell command does not define its own value for this setting. 👍 

## So, todo:
- [ ] ~~Make it possible to show the notification in a modal instead of a balloon (toggleable in settings).~~ (Edit 2022-10-21: Maybe not, as it might become redundant with _Ask after execution modal_ after [realtime stdout handling](https://github.com/Taitava/obsidian-shellcommands/discussions/64) will be implemented? See more in my other reply below.)
- [ ] Make it possible to define these settings on shell command level.

I'll create issues for these when I start working on these, but I'm not able to do it at this point, will take some time. I'll update this discussion when these are under development.

**Reply by snail123815** [reply]
Agree that **blocking** may not be worth the effort for an uncertain result.

I don't know what modal means but I suppose it is something more aberrant than a balloon that catches attention. If it is the case, I fully agree. And if the output can be in realtime shown on it, that will be more than perfect.

Thanks for the nice tool again! It is becoming a power house for me!

**Reply by Taitava** [reply]
> I don't know what modal means

A _modal_ is a big popup box that covers all normal content when it opens up. On webpages, modals are often used in image galleries to show bigger versions of images. A good example in Obsidian is the settings modal:

![kuva](https://user-images.githubusercontent.com/13002142/197235968-fe01a1da-3854-40c0-b70f-68cd1a4f826f.png)

The modal covers the editor, preventing user from interacting with it before they close the modal.

> And if the output can be in realtime shown on it, that will be more than perfect.

Now that you mentioned realtime output, it could actually make sense to perhaps postpone this feature a bit. I first thought that this could create a new type of modal, but it might become redundant if I ever get [realtime stdout handling](https://github.com/Taitava/obsidian-shellcommands/discussions/64) implemented. Currently, there is already an output modal, but it only appears **after** execution completes:
![kuva](https://user-images.githubusercontent.com/13002142/197237058-caedcae1-8a0c-45ad-9a00-3e29c882d226.png)

When output handling will be realtime, choosing [Ask after execution](https://publish.obsidian.md/shellcommands/Output+handling/Output+channel+-+Ask+after+execution)* as an output channel for a shell command could open up the modal immeditately when the shell command starts executing, thus preventing user from accessing the editor.

*) When output becomes realtime, I need to rename _Ask after execution_ to something else so that it does not contain the _after_ word. Edit 2022-12-07: I did not remember to do the renaming, but I will think about it later, if a good alternative name comes into mind.

**Reply by Taitava** [reply]
> Thanks for the nice tool again! It is becoming a power house for me!

Thank you for your kind words! ❤️

**Reply by Taitava** [reply]
> I just noticed the above function is implemented as a 'global' settings. Might be good to have it per command.

Hi @snail123815 . It's been a long time, but I'm finally beginning to implement the possibility to specify the execution notification setting on a per shell command basis. The process can be tracked in issue #388.

**Reply by Taitava** [reply]
@snail123815 if you want, you can try the setting in [`0.22.0-beta.1`](https://github.com/Taitava/obsidian-shellcommands/discussions/391). It's been open for a week now, I just didn't remember to post about it in this discussion.

**Reply by snail123815** [reply]
Hi Taitava,
Sorry for the late reply. I just had some time to test the [beta](https://github.com/Taitava/obsidian-shellcommands/discussions/391).

The settings are working fine. I tried two commands with different notification settings.

Thanks for the effort, you made the plugin better!

**Reply by Taitava** [reply]
Thank you for the feedback! ❤️

**Reply by Taitava** [reply]
## Released
This per shell command setting is now finally released in `0.22.0`.
